qabelbox/src/test/java/de/qabel/qabelbox/storage/notifications/StorageNotificationTest.java

```diff
@@ -0,0 +1,110 @@
+
+import java.util.LinkedList;
+import java.util.Queue;
+
+import de.qabel.qabelbox.BuildConfig;
+import de.qabel.qabelbox.R;
+import de.qabel.qabelbox.SimpleApplication;
+import de.qabel.qabelbox.storage.BoxUploadingFile;
+
+import static org.junit.Assert.assertEquals;
+import static org.robolectric.Shadows.shadowOf;
+
+@RunWith(RobolectricGradleTestRunner.class)
+@Config(application = SimpleApplication.class, constants = BuildConfig.class)
+public class StorageNotificationTest {
+
+    private static final String TEST_OWNER = "owner";
+    private static final String UPLOAD_PATH = "path";
+
+    private NotificationManager notificationManager;
+    private StorageNotificationManager storageNotificationManager;
+
+    @Before
+    public void setUp() {
+        notificationManager = (NotificationManager) RuntimeEnvironment.application
+                .getSystemService(Context.NOTIFICATION_SERVICE);
+        shadowOf(notificationManager).cancelAll();
+        storageNotificationManager = new StorageNotificationManager(
+                RuntimeEnvironment.application.getApplicationContext());
+    }
+
+    @Test
+    public void testSingleUpload() {
+        BoxUploadingFile uploadingFile = new BoxUploadingFile("FILE", UPLOAD_PATH, TEST_OWNER);
+        Queue<BoxUploadingFile> queue = new LinkedList<>();
+        queue.add(uploadingFile);
+        storageNotificationManager.updateUploadNotification(queue);
+        checkNotification(queue);
+
+        queue.poll();
+        storageNotificationManager.updateUploadNotification(queue);
+        checkNotification(queue);
+    }
+
+    @Test
+    public void testMultipleUpload() {
+        Queue<BoxUploadingFile> queue = new LinkedList<>();
+        queue.add(new BoxUploadingFile("FILE", UPLOAD_PATH, TEST_OWNER));
+        queue.add(new BoxUploadingFile("FILE2", UPLOAD_PATH, TEST_OWNER));
+        queue.add(new BoxUploadingFile("FILE3", UPLOAD_PATH, TEST_OWNER));
+        //3 Uploads
+        storageNotificationManager.updateUploadNotification(queue);
+        checkNotification(queue);
+
+        //2 Uploads
+        queue.poll();
+        storageNotificationManager.updateUploadNotification(queue);
+        checkNotification(queue);
+        //1 Upload
+
+        queue.poll();
+        storageNotificationManager.updateUploadNotification(queue);
+        checkNotification(queue);
+
+        //Uploads complete
+        queue.poll();
+        storageNotificationManager.updateUploadNotification(queue);
+        checkNotification(queue);
+    }
+
+    private void checkNotification(Queue<BoxUploadingFile> queue) {
+        BoxUploadingFile currentUpload = queue.peek();
+
+        ShadowNotificationManager shadowManger = shadowOf(notificationManager);
+        assertEquals(1, shadowManger.size());
+        Notification notification = shadowManger.getAllNotifications().get(0);
+        ShadowNotification shadowNotification = shadowOf(notification);
+
+        if (currentUpload != null) {
```

Comment:
There shouldn't be if statements in tests. You can restructure the test so that you don't have an uncertainty on wether "currentUpload" may be null. The test can be split in multiple smaller tests in which you always know if "currentUpload" is null. Having such a large assertion method is a code smell.

One way to restructure the code is to extract creation of the android notification in another class "StorageNotificationSender" and test that separateley. When you then test the StorageNotificationManager, you can substitute the StorageNotification sender with a mock and don't have to test the inner working of an android notification every time. The checkNotification() assertion could then become very simple and easy to read.